### ISSUE TITLE
Setting global collation for SQLite doesn't set collation for individual columns

### ISSUE THREAD

**Issue by gthvidsten** [dotnet/efcore#32051]
When setting `modelBuilder.UseCollation("NOCASE");` in `OnModelCreating()` then text columns do not get this collation.

Suggested as bug in [StackOverflow](https://stackoverflow.com/a/77283802/4353606).

## Steps to reproduce

1. Create a .NET7 console application
2. Add the `Microsoft.EntityFrameworkCore.Sqlite` version 7.0.12
3. Create the following `DbContext` and data model:
```C#
public class MyDbContext : DbContext
{
    public DbSet<MyDbTable> MyTable { get; set; }

    public string DbPath { get; }

    public MyDbContext()
    {
        // Copied from "Getting Started with EF Core" https://learn.microsoft.com/en-us/ef/core/get-started/overview/first-app
        var folder = Environment.SpecialFolder.LocalApplicationData;
        var path = Environment.GetFolderPath(folder);
        DbPath = Path.Join(path, "mydb.db");
    }

    protected override void OnConfiguring(DbContextOptionsBuilder options)
        => options.UseSqlite($"Data Source={DbPath}");


    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.UseCollation("NOCASE");
        base.OnModelCreating(modelBuilder);
    }
}

public class MyDbTable
{
    public int Id { get; set; }
    public string Title { get; set; }
}
```
4. Run `Add-Migration Initial` to create the migration
5. Run `Update-Database" to apply the migration
6. Check the `MyTable.Title` column in the database to see that it does not have the `NOCASE` collation
7. Run `Update-Database -Migration:0` to reset the database
8. Run `Remove-Migration`
9. Replace `modelBuilder.UseCollation("NOCASE");` with `modelBuilder.Entity<MyDbTable>().Property(t => t.Title).HasColumnType("TEXT COLLATE NOCASE");`
10. Run `Add-Migration Initial`
11. Run `Update-Database`
12. Check the `MyTable.Title` column to see that you can set the `NOCASE` individually

### Provider and version information

EF Core version:
Database provider: Microsoft.EntityFrameworkCore.Sqlite 7.0.12
Target framework: .NET 7.0
Operating system: Windows 11 Pro 22H2 (22621.2428)
IDE: Visual Studio Professional 2022 17.7.5

**Comment by roji** [comment]
AFAIK Sqlite does not support a database-level collation like other databases, which is why UseCollation at the modelBuilder level doesn't do anything.

However, you can use EF bulk-configuration to the collation on all string properties as follows:

```c#
protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
{
    configurationBuilder.Properties<string>().UseCollation("NOCASE");
}
```

**Comment by bricelam** [comment]
Can we enforce it with a provider-specific convention? I think setting it at the model level should take effect whether or not there is a database default. At the very least, we should warn.

**Comment by bricelam** [comment]
In other words it should *behave* more like HasDefaultSchema regardless of how it’s implemented.

**Comment by roji** [comment]
Yeah, we should be able to have a Sqlite convention that picks up the model-level collation and applies it to all textual properties... I think that makes sense...

**Comment by roji** [comment]
@joaoramalho please open a new issue with a minimal, runnable repro showing this.

**Comment by zang227** [comment]
> AFAIK Sqlite does not support a database-level collation like other databases, which is why UseCollation at the modelBuilder level doesn't do anything.
> 
> However, you can use EF bulk-configuration to the collation on all string properties as follows:
> 
> protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
> {
>     configurationBuilder.Properties<string>().UseCollation("NOCASE");
> }

We tried this in our project but it doesn't seem to be working for us. 


```c#
        protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
        {
            base.ConfigureConventions(configurationBuilder);

            configurationBuilder.Properties<string>().UseCollation("NOCASE");
        }
```



We use linq to query

```c#
            inventoryItems = (from i in inventoryItems
                              where (
                                i.Code.Contains(SearchTerm)
                                ||
                                i.Description.Contains(SearchTerm)
                                ||
                                i.DigitalId.Contains(SearchTerm)
                              )
                              select i);
```

But it doesn't seem to affect the query at all and no data is returned unless SearchTerm matches the case of i.Code

Was the solution you proposed verified to be working?

**Comment by bricelam** [comment]
@zang227 i wonder if `instr` ignores the collation…

**Comment by zang227** [comment]
> [@zang227](https://github.com/zang227) i wonder if `instr` ignores the collation…

Yeah I think it might ignore it.

 Using 

`EF.Functions.Collate(i.Code, "NOCASE").Contains(SearchTerm));`


We were able to get NOCASE to show in the query

`FROM "InventoryItems" AS "i"

WHERE instr("i"."Code" COLLATE NOCASE, @__SearchTerm_1) > 0 OR ("i"."Description" IS NOT NULL AND instr("i"."Description", @__SearchTerm_1) > 0) OR ("i"."DigitalId" IS NOT NULL AND instr("i"."DigitalId", @__SearchTerm_1) > 0)
 `


But this still failed to yield the data we were expecting.

I also tried 

`EF.Functions.Collate(i.Code, "NOCASE").Contains(EF.Functions.Collate(SearchTerm, "NOCASE")));`

But this didn't work either.

**Comment by bricelam** [comment]
A less-than-ideal solution would be to convert both sides to upper-case:

```cs
var searchTerm = SearchTerm.ToUpper();
inventoryItems = from i in inventoryItems
                 where i.Code.ToUpper().Contains(searchTerm)
                     || i.Description.ToUpper().Contains(searchTerm)
                     || i.DigitalId.ToUpper().Contains(searchTerm)
                 select i;
```

**Comment by zang227** [comment]
> A less-than-ideal solution would be to convert both sides to upper-case:
> 
> var searchTerm = SearchTerm.ToUpper();
> inventoryItems = from i in inventoryItems
>                  where i.Code.ToUpper().Contains(searchTerm)
>                      || i.Description.ToUpper().Contains(searchTerm)
>                      || i.DigitalId.ToUpper().Contains(searchTerm)
>                  select i;

Yeah we'd like to avoid this if possible due to the number of queries already existing that we'd have to update and having all our devs remember to do it every time.

**Comment by bricelam** [comment]
Replace the `instr` implementation with your own case-insensitive version:

```cs
sqliteConnection.CreateFunction(
    "instr",
    (string x, string y) => x.Contains(y, StringComparison.OrdinalIgnoreCase),
    isDeterministic: true);
```

**Comment by zang227** [comment]
> Replace the `instr` implementation with your own case-insensitive version:
> 
> sqliteConnection.CreateFunction(
>     "instr",
>     (string x, string y) => x.Contains(y, StringComparison.OrdinalIgnoreCase),
>     isDeterministic: true);

This worked. Thank you!
